sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala

```diff
@@ -293,27 +293,27 @@ object SQLConf {
       .bytesConf(ByteUnit.BYTE)
       .createWithDefault(64 * 1024 * 1024)
 
-  val RUNTIME_REOPTIMIZATION_ENABLED =
-    buildConf("spark.sql.runtime.reoptimization.enabled")
-      .doc("When true, enable runtime query re-optimization.")
-      .booleanConf
-      .createWithDefault(false)
-
   val ADAPTIVE_EXECUTION_ENABLED = buildConf("spark.sql.adaptive.enabled")
```

Comment:
This is an umbrella config, let's create a new config for `ReduceNumShufflePartitions`. How about `spark.sql.adaptive.reducePostShufflePartitions.enabled`?

Reply:
+1. Being a bit pedantic here, but let's try to do orthogonal changes like removing the `RUNTIME_REOPTIMIZATION_ENABLED` flag in separate PRs. That makes the cognitive load of reviewing lower, and it also improves git blame.

Reply:
Agree to add a config for `ReduceNumShufflePartitions`. I was removing `RUNTIME_REOPTIMIZATION_ENABLED` because `RUNTIME_REOPTIMIZATION_ENABLED` and `ADAPTIVE_EXECUTION_ENABLED` look like the same config now. I didn't see a way to enable `ReduceNumShufflePartitions` but disable runtime reoptimization.

Reply:
The argument of not changing `RUNTIME_REOPTIMIZATION_ENABLED` to `ADAPTIVE_EXECUTION_ENABLED` is just to keep this PR independent and smaller.

Whatever name it is, we should have an umbrella flag that turns everything off. We could add another flag, though, just to turn off the SMJ -> BHJ runtime optimization, although we might have to do it in a slightly trickier way given that this opt comes naturally with the framework itself.

Reply:
All right, I will use `RUNTIME_REOPTIMIZATION_ENABLED` as the umbrella flag in this PR.